src/service/standard-actions-impl.js

```diff
@@ -172,11 +173,26 @@ export class StandardActions {
       return null;
     }
     const node = invocation.node;
-    const win = (node.ownerDocument || node).defaultView;
-    const url = invocation.args['url'];
-    const requestedBy = `AMP.${invocation.method}`;
-    Services.navigationForDoc(this.ampdoc).navigateTo(win, url, requestedBy);
-    return null;
+    // Some components have additional constraints on allowing navigation.
+    let permission = Promise.resolve();
+    if (startsWith(node.tagName, 'AMP-')) {
+      permission = node.getImpl().then(impl => {
+        if (typeof impl.navigationError == 'function') {
+          const error = impl.navigationError();
+          if (error) {
+            throw error;
+          }
+        }
+      });
+    }
+    return permission.then(() => {
+      const win = (node.ownerDocument || node).defaultView;
+      const url = invocation.args['url'];
+      const requestedBy = `AMP.${invocation.method}`;
+      Services.navigationForDoc(this.ampdoc).navigateTo(win, url, requestedBy);
```

Comment:
I'm curious if this will have issues with deferred action. Ie, user clicks, postmessage comes out of iframe telling it to navigate. If we defer the navigate from the click, will the browser still let the navigate happen?

Reply:
I know that some ad blockers will follow the code path for click events on `<a>` tags but can't traverse `postMessage` boundary. I haven't heard the same for browsers in general.

Publishers can always just perform navigation in the iframe directly if `allow-top-navigation` is set. I'm actually not sure why publishers would use `AMP.navigateTo` in this case, but we should still do this check regardless.